// @metaplex-foundation/umi-serializer-beet/dist/esm/struct.mjs
import { mergeBytes } from '@metaplex-foundation/umi';
import { sumSerializerSizes } from './sumSerializerSizes.mjs';

function struct(fields, options = {}) {
  const fieldDescriptions = fields.map(([name, serializer]) => `${String(name)}: ${serializer.description}`).join(', ');
  return {
    description: options.description ?? `struct(${fieldDescriptions})`,
    fixedSize: sumSerializerSizes(fields.map(([, field]) => field.fixedSize)),
    maxSize: sumSerializerSizes(fields.map(([, field]) => field.maxSize)),
    serialize: struct => {
      const fieldBytes = fields.map(([key, serializer]) => serializer.serialize(struct[key]));
      return mergeBytes(fieldBytes);
    },
    deserialize: (bytes, offset = 0) => {
      const struct = {};
      fields.forEach(([key, serializer]) => {
        const [value, newOffset] = serializer.deserialize(bytes, offset);
        offset = newOffset;
        struct[key] = value;
      });
      return [struct, offset];
    }
  };
}

export { struct };
