// @metaplex-foundation/umi-serializer-beet/dist/esm/tuple.mjs
import { mergeBytes } from '@metaplex-foundation/umi';
import { BeetSerializerError } from './errors.mjs';
import { sumSerializerSizes } from './sumSerializerSizes.mjs';

function tuple(items, options = {}) {
  const itemDescriptions = items.map(item => item.description).join(', ');
  return {
    description: options.description ?? `tuple(${itemDescriptions})`,
    fixedSize: sumSerializerSizes(items.map(item => item.fixedSize)),
    maxSize: sumSerializerSizes(items.map(item => item.maxSize)),
    serialize: value => {
      if (value.length !== items.length) {
        throw new BeetSerializerError(`Expected tuple to have ${items.length} items but got ${value.length}.`);
      }
      return mergeBytes(items.map((item, index) => item.serialize(value[index])));
    },
    deserialize: (bytes, offset = 0) => {
      const values = [];
      items.forEach(serializer => {
        const [newValue, newOffset] = serializer.deserialize(bytes, offset);
        values.push(newValue);
        offset = newOffset;
      });
      return [values, offset];
    }
  };
}

export { tuple };
