// @metaplex-foundation/umi-serializer-beet/dist/cjs/array.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var umi = require('@metaplex-foundation/umi');
var errors = require('./errors.cjs');
var getResolvedSize = require('./getResolvedSize.cjs');
var getSizeDescription = require('./getSizeDescription.cjs');
var getSizeFromChildren = require('./getSizeFromChildren.cjs');
var getSizePrefix = require('./getSizePrefix.cjs');
var numbers = require('./numbers.cjs');

function array(item, options = {}) {
  const size = options.size ?? numbers.u32();
  if (size === 'remainder' && item.fixedSize === null) {
    throw new errors.BeetSerializerError('Serializers of "remainder" size must have fixed-size items.');
  }
  return {
    description: options.description ?? `array(${item.description}; ${getSizeDescription.getSizeDescription(size)})`,
    fixedSize: getSizeFromChildren.getSizeFromChildren(size, [item.fixedSize]),
    maxSize: getSizeFromChildren.getSizeFromChildren(size, [item.maxSize]),
    serialize: value => {
      if (typeof size === 'number' && value.length !== size) {
        throw new errors.BeetSerializerError(`Expected array to have ${size} items but got ${value.length}.`);
      }
      return umi.mergeBytes([getSizePrefix.getSizePrefix(size, value.length), ...value.map(v => item.serialize(v))]);
    },
    deserialize: (bytes, offset = 0) => {
      if (typeof size === 'object' && bytes.slice(offset).length === 0) {
        throw new errors.DeserializingEmptyBufferError('array', []);
      }
      const [resolvedSize, newOffset] = getResolvedSize.getResolvedSize(size, [item.fixedSize], bytes, offset);
      offset = newOffset;
      const values = [];
      for (let i = 0; i < resolvedSize; i += 1) {
        const [value, newOffset] = item.deserialize(bytes, offset);
        values.push(value);
        offset = newOffset;
      }
      return [values, offset];
    }
  };
}

exports.array = array;
