// @metaplex-foundation/umi-serializer-beet/dist/cjs/tuple.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var umi = require('@metaplex-foundation/umi');
var errors = require('./errors.cjs');
var sumSerializerSizes = require('./sumSerializerSizes.cjs');

function tuple(items, options = {}) {
  const itemDescriptions = items.map(item => item.description).join(', ');
  return {
    description: options.description ?? `tuple(${itemDescriptions})`,
    fixedSize: sumSerializerSizes.sumSerializerSizes(items.map(item => item.fixedSize)),
    maxSize: sumSerializerSizes.sumSerializerSizes(items.map(item => item.maxSize)),
    serialize: value => {
      if (value.length !== items.length) {
        throw new errors.BeetSerializerError(`Expected tuple to have ${items.length} items but got ${value.length}.`);
      }
      return umi.mergeBytes(items.map((item, index) => item.serialize(value[index])));
    },
    deserialize: (bytes, offset = 0) => {
      const values = [];
      items.forEach(serializer => {
        const [newValue, newOffset] = serializer.deserialize(bytes, offset);
        values.push(newValue);
        offset = newOffset;
      });
      return [values, offset];
    }
  };
}

exports.tuple = tuple;
